storage/migrations/18_refactor_dead_reckoning.up.sql

```diff
@@ -0,0 +1,108 @@
+BEGIN;
+
+-------------------------------------
+-- Convenience functions for converting between eth-style and oasis-style addresses
+-- via lookups in the address_preimages table.
+
+-- Convenience function for retrieving the ethereum address associated with a given oasis address.
+CREATE OR REPLACE FUNCTION eth_preimage(addr oasis_addr)
+RETURNS BYTEA
+LANGUAGE plpgsql
+AS $$
+BEGIN
+DECLARE
+    result BYTEA;
+BEGIN
+    SELECT pre.address_data
+    INTO result
+    FROM chain.address_preimages pre
+    WHERE pre.address = addr
+    AND pre.context_identifier = 'oasis-runtime-sdk/address: secp256k1eth'
+    AND pre.context_version = 0;
+    RETURN result;
+END;
+END;
+$$;
+
+CREATE OR REPLACE FUNCTION derive_oasis_addr(eth_addr bytea)
+RETURNS TEXT
+LANGUAGE plpgsql
+AS $$
+BEGIN
+DECLARE
+    result TEXT;
+BEGIN
+    -- Check if the bytea data is exactly 20 bytes in length; this prevents accidentally accepting strings that are hex-encodings of the actual address, or similar.
+    IF length(eth_addr) <> 20 THEN
+        RAISE EXCEPTION 'Input address must be a bytea and exactly 20 bytes in length';
+    END IF;
+
+    -- Look up the oasis-style address derived from evs.body.address.
+    -- The derivation is just a keccak hash and we could theoretically compute it instead of looking it up,
+    -- but the right hash function will only be easily available in postgres once openssl 3.2 is released
+    -- (see https://github.com/openssl/openssl/issues/19304) and bundled into a postgres docker image, probably
+    -- in early 2024.
+    SELECT pre.address
+    INTO result
+    FROM chain.address_preimages pre
+    WHERE pre.address_data = eth_addr
+    AND pre.context_identifier = 'oasis-runtime-sdk/address: secp256k1eth'
+    AND pre.context_version = 0;
+    RETURN result;
+END;
+END;
+$$;
+
+
+-------------------------------------
+-- Moves the chain.evm_contracts.gas_used column to chain.runtime_accounts.gas_for_calling
+
+ALTER TABLE chain.runtime_accounts
+    -- Total gas used by all txs addressed to this account. Primarily meaningful for accounts that are contracts.
+    ADD COLUMN gas_for_calling UINT63 NOT NULL DEFAULT 0;
+
+UPDATE chain.runtime_accounts as ra
+    SET gas_for_calling = c.gas_used
+    FROM chain.evm_contracts c
+    WHERE c.runtime = ra.runtime AND
+          c.contract_address = ra.address;
+
+ALTER TABLE chain.evm_contracts
+    DROP COLUMN gas_used;
```

Comment:
There are currently on non-(evm-contract) addresses that we'd want to track gas_for_calling for. If we add support for cipher, there will be (non-evm)-contract addresses that we might well want to include in this reckoning.

I could also see a potential argument for why we'd want to track this for non-contracts: If people spend a lot of gas for sending money to an account, that's a noteworthy account.

All of that said, the main reason to track this outside the `evm_contracts` table is to avoid the problem of how/where to do dead reckoning for an address before we know for sure that it's a contract (and thus before it has an entry in `evm_contracts`). Well, and also Cipher or any future runtimes; those contracts won't be in `evm_contracts`.

I think it's fair to have this here; it's a property of every account, it's just that it's the most interesting for contract accounts. I hope the comment in line 61 clarifies enough.